File: go/connector/notion_blocks.go
// SPDX-License-Identifier: Apache-2.0

package connector

import (
	"encoding/json"
	"fmt"
	"strings"
)

// notionSearchResponse models the Notion search API response.
type notionSearchResponse struct {
	Results    []json.RawMessage `json:"results"`
	NextCursor string            `json:"next_cursor"`
	HasMore    bool              `json:"has_more"`
}

// notionDatabaseQueryResponse models the database query response.
type notionDatabaseQueryResponse struct {
	Results    []json.RawMessage `json:"results"`
	NextCursor string            `json:"next_cursor"`
	HasMore    bool              `json:"has_more"`
}

// notionBlockChildrenResponse models the block children response.
type notionBlockChildrenResponse struct {
	Results    []notionBlock `json:"results"`
	NextCursor string       `json:"next_cursor"`
	HasMore    bool         `json:"has_more"`
}

// notionBlock represents a single Notion block.
type notionBlock struct {
	ID          string          `json:"id"`
	Type        string          `json:"type"`
	HasChildren bool            `json:"has_children"`
	RawData     json.RawMessage `json:"-"`
}

// UnmarshalJSON implements custom unmarshalling for notionBlock to
// capture the full raw data alongside parsed fields.
func (b *notionBlock) UnmarshalJSON(data []byte) error {
	type alias notionBlock
	var a alias
	if err := json.Unmarshal(data, &a); err != nil {
		return err
	}
	a.RawData = data
	*b = notionBlock(a)
	return nil
}

// notionRichText represents a Notion rich text object.
type notionRichText struct {
	Type        string             `json:"type"`
	PlainText   string             `json:"plain_text"`
	Href        string             `json:"href"`
	Annotations notionAnnotations  `json:"annotations"`
}

// notionAnnotations contains formatting flags for rich text.
type notionAnnotations struct {
	Bold          bool   `json:"bold"`
	Italic        bool   `json:"italic"`
	Strikethrough bool   `json:"strikethrough"`
	Underline     bool   `json:"underline"`
	Code          bool   `json:"code"`
	Colour        string `json:"color"`
}

// notionListBlockTypes is the set of block types considered list items
// for indentation purposes. Module-level constant avoids per-call
// allocation.
var notionListBlockTypes = map[string]struct{}{
	"bulleted_list_item": {},
	"numbered_list_item": {},
	"to_do":              {},
	"toggle":             {},
}

// isListBlock returns true if the block type is a list item.
func isListBlock(blockType string) bool {
	_, ok := notionListBlockTypes[blockType]
	return ok
}

// blockToMarkdown converts a single Notion block to its markdown
// representation.
func (c *NotionConnector) blockToMarkdown(block notionBlock) string {
	// Skip blocks excluded by the block type filter.
	if len(c.config.BlockTypeFilter) > 0 {
		if _, allowed := c.config.BlockTypeFilter[block.Type]; !allowed {
			return ""
		}
	}

	var raw map[string]json.RawMessage
	if err := json.Unmarshal(block.RawData, &raw); err != nil {
		return ""
	}

	blockData, ok := raw[block.Type]
	if !ok {
		return ""
	}

	var typed struct {
		RichText   []notionRichText `json:"rich_text"`
		Caption    []notionRichText `json:"caption"`
		Language   string           `json:"language"`
		URL        string           `json:"url"`
		Checked    bool             `json:"checked"`
		Expression string           `json:"expression"`
	}
	if err := json.Unmarshal(blockData, &typed); err != nil {
		return ""
	}

	text := renderRichText(typed.RichText)

	blockRenderers := map[string]func() string{
		"paragraph":          func() string { return text + "\n" },
		"heading_1":          func() string { return "# " + text + "\n" },
		"heading_2":          func() string { return "## " + text + "\n" },
		"heading_3":          func() string { return "### " + text + "\n" },
		"bulleted_list_item": func() string { return "- " + text },
		"numbered_list_item": func() string { return "1. " + text },
		"to_do": func() string {
			marker := "[ ]"
			if typed.Checked {
				marker = "[x]"
			}
			return "- " + marker + " " + text
		},
		"toggle":    func() string { return "- " + text },
		"quote":     func() string { return "> " + text + "\n" },
		"callout":   func() string { return "> " + text + "\n" },
		"divider":   func() string { return "---\n" },
		"code": func() string {
			lang := typed.Language
			return "```" + lang + "\n" + text + "\n```\n"
		},
		"equation": func() string {
			return "$$\n" + typed.Expression + "\n$$\n"
		},
		"image": func() string {
			caption := renderRichText(typed.Caption)
			url := extractFileURL(blockData)
			if caption != "" {
				return fmt.Sprintf("![%s](%s)\n", caption, url)
			}
			return fmt.Sprintf("![](%s)\n", url)
		},
		"file": func() string {
			url := extractFileURL(blockData)
			caption := renderRichText(typed.Caption)
			if caption == "" {
				caption = "file"
			}
			return fmt.Sprintf("[%s](%s)\n", caption, url)
		},
		"bookmark": func() string {
			caption := renderRichText(typed.Caption)
			if typed.URL != "" {
				if caption != "" {
					return fmt.Sprintf("[%s](%s)\n", caption, typed.URL)
				}
				return typed.URL + "\n"
			}
			return ""
		},
		"embed": func() string {
			if typed.URL != "" {
				return typed.URL + "\n"
			}
			return ""
		},
		"table_of_contents": func() string { return "" },
		"breadcrumb":        func() string { return "" },
		"column_list":       func() string { return "" },
		"column":            func() string { return "" },
		"child_page":        func() string { return "" },
		"child_database":    func() string { return "" },
		"synced_block":      func() string { return "" },
		"link_preview": func() string {
			if typed.URL != "" {
				return typed.URL + "\n"
			}
			return ""
		},
		"table":     func() string { return "" },
		"table_row": func() string { return renderTableRow(blockData) },
	}

	renderer, found := blockRenderers[block.Type]
	if !found {
		return text
	}
	return renderer()
}

// renderRichText converts a slice of Notion rich text objects to a
// markdown string with formatting annotations.
func renderRichText(texts []notionRichText) string {
	var builder strings.Builder
	for _, t := range texts {
		text := t.PlainText

		if t.Annotations.Code {
			text = "`" + text + "`"
		}
		if t.Annotations.Bold {
			text = "**" + text + "**"
		}
		if t.Annotations.Italic {
			text = "*" + text + "*"
		}
		if t.Annotations.Strikethrough {
			text = "~~" + text + "~~"
		}

		if t.Href != "" {
			text = "[" + text + "](" + t.Href + ")"
		}

		builder.WriteString(text)
	}
	return builder.String()
}

// renderPlainRichText extracts plain text from a rich text array
// without formatting.
func renderPlainRichText(texts []notionRichText) string {
	var parts []string
	for _, t := range texts {
		parts = append(parts, t.PlainText)
	}
	return strings.Join(parts, "")
}

// extractFileURL pulls the URL from a file-type block's JSON data.
func extractFileURL(data json.RawMessage) string {
	var fileBlock struct {
		External struct {
			URL string `json:"url"`
		} `json:"external"`
		File struct {
			URL string `json:"url"`
		} `json:"file"`
	}
	if err := json.Unmarshal(data, &fileBlock); err != nil {
		return ""
	}
	if fileBlock.File.URL != "" {
		return fileBlock.File.URL
	}
	return fileBlock.External.URL
}

// renderTableRow converts a table_row block to a markdown table
// row.
func renderTableRow(data json.RawMessage) string {
	var row struct {
		Cells [][]notionRichText `json:"cells"`
	}
	if err := json.Unmarshal(data, &row); err != nil {
		return ""
	}

	cells := make([]string, 0, len(row.Cells))
	for _, cell := range row.Cells {
		cells = append(cells, renderRichText(cell))
	}

	return "| " + strings.Join(cells, " | ") + " |"
}

File: sdks/ts/memory/src/connector/notion-blocks.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * Block rendering helpers for the Notion connector. Converts Notion
 * block objects to markdown representation and handles rich text
 * formatting.
 */

// ---------- Types ----------

/**
 * Notion rich text object.
 */
export type NotionRichText = {
  readonly type: string
  readonly plain_text: string
  readonly href?: string
  readonly annotations?: NotionAnnotations
}

/**
 * Notion text formatting annotations.
 */
type NotionAnnotations = {
  readonly bold?: boolean
  readonly italic?: boolean
  readonly strikethrough?: boolean
  readonly underline?: boolean
  readonly code?: boolean
  readonly color?: string
}

/**
 * Notion block object from the blocks API.
 */
export type NotionBlock = {
  readonly id: string
  readonly type: string
  readonly has_children: boolean
  readonly [key: string]: unknown
}

// ---------- Module-level constants ----------

/**
 * Set of block types considered list items for indentation.
 * Module-level constant avoids per-call allocation.
 */
const LIST_BLOCK_TYPES = new Set([
  'bulleted_list_item',
  'numbered_list_item',
  'to_do',
  'toggle',
])

// ---------- Rich text helpers ----------

/**
 * Renders a Notion rich text array to markdown with formatting.
 */
export const renderRichText = (texts: readonly NotionRichText[]): string => {
  const parts: string[] = []

  for (const t of texts) {
    let text = t.plain_text

    if (t.annotations?.code === true) {
      text = '`' + text + '`'
    }
    if (t.annotations?.bold === true) {
      text = '**' + text + '**'
    }
    if (t.annotations?.italic === true) {
      text = '*' + text + '*'
    }
    if (t.annotations?.strikethrough === true) {
      text = '~~' + text + '~~'
    }

    if (t.href !== undefined && t.href !== '') {
      text = `[${text}](${t.href})`
    }

    parts.push(text)
  }

  return parts.join('')
}

/**
 * Extracts plain text from rich text without formatting.
 */
export const renderPlainRichText = (
  texts: readonly NotionRichText[],
): string => texts.map((t) => t.plain_text).join('')

// ---------- Block rendering ----------

/**
 * Converts a single Notion block to markdown. Blocks whose type is
 * not in the allowedTypes filter (when provided) are skipped.
 */
export const blockToMarkdown = (
  block: NotionBlock,
  allowedTypes?: ReadonlySet<string>,
): string => {
  // Apply block type filter when configured.
  if (allowedTypes !== undefined && !allowedTypes.has(block.type)) {
    return ''
  }

  const blockData = block[block.type] as Record<string, unknown> | undefined
  if (blockData === undefined) return ''

  const richText = (blockData.rich_text ?? []) as readonly NotionRichText[]
  const text = renderRichText(richText)
  const caption = (blockData.caption ?? []) as readonly NotionRichText[]

  const renderers: Readonly<Record<string, () => string>> = {
    paragraph: () => `${text}\n`,
    heading_1: () => `# ${text}\n`,
    heading_2: () => `## ${text}\n`,
    heading_3: () => `### ${text}\n`,
    bulleted_list_item: () => `- ${text}`,
    numbered_list_item: () => `1. ${text}`,
    to_do: () => {
      const checked = blockData.checked === true
      const marker = checked ? '[x]' : '[ ]'
      return `- ${marker} ${text}`
    },
    toggle: () => `- ${text}`,
    quote: () => `> ${text}\n`,
    callout: () => `> ${text}\n`,
    divider: () => '---\n',
    code: () => {
      const lang = (blockData.language as string) ?? ''
      return '```' + lang + '\n' + text + '\n```\n'
    },
    equation: () => {
      const expression = (blockData.expression as string) ?? ''
      return `$$\n${expression}\n$$\n`
    },
    image: () => {
      const capText = renderRichText(caption)
      const url = extractFileUrl(blockData)
      return capText !== ''
        ? `![${capText}](${url})\n`
        : `![](${url})\n`
    },
    file: () => {
      const url = extractFileUrl(blockData)
      const capText = renderRichText(caption)
      const label = capText !== '' ? capText : 'file'
      return `[${label}](${url})\n`
    },
    bookmark: () => {
      const capText = renderRichText(caption)
      const url = blockData.url as string | undefined
      if (url !== undefined) {
        return capText !== '' ? `[${capText}](${url})\n` : `${url}\n`
      }
      return ''
    },
    embed: () => {
      const url = blockData.url as string | undefined
      return url !== undefined ? `${url}\n` : ''
    },
    table_of_contents: () => '',
    breadcrumb: () => '',
    column_list: () => '',
    column: () => '',
    child_page: () => '',
    child_database: () => '',
    synced_block: () => '',
    link_preview: () => {
      const url = blockData.url as string | undefined
      return url !== undefined ? `${url}\n` : ''
    },
    table: () => '',
    table_row: () => {
      const cells = (blockData.cells ?? []) as readonly (readonly NotionRichText[])[]
      const rendered = cells.map((cell) => renderRichText(cell))
      return `| ${rendered.join(' | ')} |`
    },
  }

  const renderer = renderers[block.type]
  return renderer !== undefined ? renderer() : text
}

/**
 * Checks whether a block type is a list item.
 */
export const isListBlock = (blockType: string): boolean =>
  LIST_BLOCK_TYPES.has(blockType)

/**
 * Extracts a file URL from a file-type block's data.
 */
const extractFileUrl = (data: Record<string, unknown>): string => {
  const file = data.file as { url?: string } | undefined
  const external = data.external as { url?: string } | undefined
  return file?.url ?? external?.url ?? ''
}
